### PR TITLE
Feature/trie

### PR DESCRIPTION
**Pull request by pguzek**

### CODE REVIEW

**Comment by pguzek** [comment]
# merry xmas

**Comment by pguzek** [comment]
bump

File: src/DataStructure/Trie/TrieNode.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        return true
+    }
+
+    remove(key: string): boolean {
+        const found = this.find(key)
+        if (!isUndefined(found)) {
+            if (found.children.length === 1) {
+                found.collapse()
+
```
Comment (TheByteBits) [inline src/DataStructure/Trie/TrieNode.ts]: Extra whitespace for no good reason

File: src/DataStructure/Trie/TrieNode.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            } else if (found.children.length > 1) {
+                found.value = undefined
+
```
Comment (TheByteBits) [inline src/DataStructure/Trie/TrieNode.ts]: Extra whitespace for no good reason

File: src/DataStructure/Trie/TrieNode.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import reduce from "List/reduce/reduce"
+
+export default class TrieNode<T> {
+
+    keypart: string
+    value: T
+    parent: TrieNode<T>
+    children: TrieNode<T>[] = []
+
+    constructor(value: T, keypart: string, parent: TrieNode<T> = undefined) {
```
Comment (TheByteBits) [inline src/DataStructure/Trie/TrieNode.ts]: Would be better if you would call it explicitly as an optional parameter instead.

constructor(value: T, keypart: string, parent?: TrieNode<T>)

File: src/DataStructure/Trie/Trie.spec.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        })
+        it('should have zero iteration steps', () => {
+            const trie = new Trie<Number>()
+            let i = 0
+            for(let x of trie) {
+                i++
+            }
+            expect(i).toEqual(0)
+        })
+        it('should add element correcly', () => {
```
Comment (TheByteBits) [inline src/DataStructure/Trie/Trie.spec.ts]: There's a typo here: "correcly"

**Comment by pguzek** [comment]
all requested changes have been added

**Pull request merged**